### DISCUSSION TITLE
Protection enchantment nerf

### DISCUSSION THREAD

**Discussion by ACGaming** [ACGaming/UniversalTweaks#274] [Tweak Suggestions]
- Make protection enchantment percentages configurable

From abadhaiku (https://discord.com/channels/932680710815350834/1091539781722320916/1145409281185820784):

Ok looks like it sums those values up to a max of 20, which is then divided by 25, for a max of 75% damage reduction.  If you want to tweak those values more effectively, you need to do the following:
1. Pick some scale `n` to increase the precision by,
2. multiply the `25.0f` and `20.0F` in `net.minecraft.util.CombatRules#getDamageAfterMagicAbsorb` by your scale,
3. mix into `EnchantmentProtection#calcModifierDamage` and apply your changes to the vanilla values, and 
4. redirect `net.minecraft.enchantment.EnchantmentHelper.ModifierDamage#calculateModifier > calcModiferDamage` to multiply it by the scale if the enchantment isn‘t vanilla protection, so the total percentage stays the same for protection enchantments that don‘t use the modifiable enchantment API you‘re creating

idk why they implemented it the way they did. enchantmenthelper is queued to use floats, but the protection values use int.
honestly it might even be easier to just overwrite the method and use your own protection implementation with `float` values.
I remember something on the Mixin wiki about the JVM allowing methods that otherwise have the same signature but have different return values. I wonder if you could migrate the `ModifierDamage.damageModifer` type to `float` instead of `int` and change `EnchantmentProtection#calcModiferDamage` to return a `float` too, since the calculation in `ModifierDamage#calculateModifier` uses a `+` operator which works with any primitive number type.
Might just be easier to just inject your own implementation into `EntityLivingBase` and make it read non-vanilla enchantments normally

**Comment by ByThePowerOfScience** [comment]
Haiku here.
Analyzing the code, I don't see any usage errors that would come up if we were to just change the variable type of `ModiferDamage.damageModifer`, `EntityLivingBase#applyPotionDamageCalculations.k`, and the return value of `EnchantmentHelper#getEnchantmentModifierDamage` from `int` to `float`.  You definitely can't use Mixin to do that, but ASM can do it for sure by modifying the bytecode and LVT directly.
